waku/v2/peermanager/rtt_cache.go

```diff
@@ -0,0 +1,256 @@
+package peermanager
+
+import (
+	"container/heap"
+	"context"
+	"math/rand"
+	"sync"
+	"time"
+
+	"github.com/libp2p/go-libp2p/core/host"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/libp2p/go-libp2p/p2p/discovery/backoff"
+	"github.com/libp2p/go-libp2p/p2p/protocol/ping"
+	"github.com/waku-org/go-waku/logging"
+	"github.com/waku-org/go-waku/waku/v2/timesource"
+	"go.uber.org/zap"
+)
+
+type PeerRTT struct {
+	pingQueueIndex int
+
+	PeerID  peer.ID
+	PingRTT time.Duration
+
+	bkf              backoff.BackoffStrategy
+	nextVerification time.Time
+	timesource       timesource.Timesource
+}
+
+func (pr *PeerRTT) IsStale() bool {
+	return pr.nextVerification.Before(pr.timesource.Now())
+}
+
+type RTTCache struct {
+	sync.RWMutex
+
+	host host.Host
+
+	peers     map[peer.ID]*PeerRTT
+	pingQueue PingQueue
+
+	verificationInterval time.Duration
+
+	timesource timesource.Timesource
+	logger     *zap.Logger
+}
+
+func NewRTTCache(timesource timesource.Timesource, logger *zap.Logger) *RTTCache {
+	return &RTTCache{
+		timesource: timesource,
+		logger:     logger.Named("ping-manager"),
+		peers:      make(map[peer.ID]*PeerRTT),
+	}
+}
+
+func (r *RTTCache) SetHost(h host.Host) {
+	r.host = h
+}
+
+func (r *RTTCache) Start(ctx context.Context) {
+	go r.start(ctx, 10*time.Second)
+}
+
+func (r *RTTCache) pingPeers(ctx context.Context) {
+	r.RLock()
+	var toPing peer.IDSlice
+	if len(r.pingQueue) == 0 {
+		toPing = r.host.Peerstore().Peers()
```

Comment:
Shouldn't we only use connectedPeers , because otherwise the ping may open a connection to the peer un-necessarily.

Reply:
But aren't we're interested in pinging all the peers to know their reply time regardless of them being connected or not?

Reply:
Yeah, but we keep discovered peers in the store in general ir-respective of whether we connect to them or not. Because we keep pruning the connections for relay if we end up connecting to more, but at the same time they still exist in the peerStore for future use, in case a new shard subscription occurs or the peers we are connected to disconnect from us. 
Hence, i was wondering if we can only do this ping for peers that we are already connected to. Although, if the application wants to select new peers for a service, then we may not have peers with best RTT to select. 
Wondering how we can get around that.

And at the same time, it will not be a good idea to keep connecting to peers just to get their RTT and not use the connections.

Reply:
I wonder if we should use instead an ICMP Ping?

Reply:
> I wonder if we should use instead an ICMP Ping?

That is not a bad idea, but generally ICMP ping also is blocked by firewalls so we can't rely on it.